shiny/ui/_theme_brand.py

```diff
@@ -0,0 +1,621 @@
+        self.functions: str | None = self._validate_str(functions, "functions")
+        self.defaults: dict[str, YamlScalarType] | None = self._validate_defaults(
+            defaults
+        )
+        self.mixins: str | None = self._validate_str(mixins, "mixins")
+        self.rules: str | None = self._validate_str(rules, "rules")
+
+    def _validate_str(self, x: Any, param: str) -> str | None:
+        if x is None or isinstance(x, str):
+            return x
+
+        raise ValueError(
+            f"Invalid brand `{self.path}.{param}`. Must be a string or empty."
+        )
+
+    def _validate_defaults(self, x: Any) -> dict[str, YamlScalarType] | None:
+        if x is None:
+            return None
+
+        path = self.path
+        if path == "defaults.shiny.theme":
+            path += ".defaults"
+
+        if not isinstance(x, dict):
+            raise ValueError(f"Invalid brand `{path}`, must be a dictionary.")
+
+        y: dict[Any, Any] = x
+
+        if not all([isinstance(k, str) for k in y.keys()]):
+            raise ValueError(f"Invalid brand `{path}`, all keys must be strings.")
+
+        if not all(
+            [v is None or isinstance(v, (str, int, float, bool)) for v in y.values()]
+        ):
+            raise ValueError(f"Invalid brand `{path}`, all values must be scalar.")
+
+        res: dict[str, YamlScalarType] = y
+        return res
+
+
+class BrandBootstrapConfig:
+    """Convenience class for storing Bootstrap defaults from a brand instance"""
+
+    def __init__(
+        self,
+        version: Any = v_bootstrap,
+        preset: str | None = None,
+        functions: str | None = None,
+        defaults: dict[str, YamlScalarType] | None = None,
+        mixins: str | None = None,
+        rules: str | None = None,
+    ):
+        if not isinstance(version, (str, int)):
+            raise ValueError(
+                f"Bootstrap version must be a string or integer, not {version!r}."
+            )
+
+        v_major = str(version).split(".")[0]
+        bs_major = str(v_bootstrap).split(".")[0]
+
+        if v_major != bs_major:
+            # TODO (bootstrap-update): Assumes Shiny ships one version of Bootstrap
+            warnings.warn(
+                f"Shiny does not current support Bootstrap version {v_major}. "
+                f"Using Bootstrap v{bs_major} instead.",
+                stacklevel=4,
+            )
+            v_major = bs_major
+
+        self.version = v_major
+        self.preset = preset
+        self.functions = functions
+        self.defaults = defaults
+        self.mixins = mixins
+        self.rules = rules
+
+    @classmethod
+    def from_brand(cls, brand: Brand):
+        if not brand.defaults:
+            return cls(version=v_bootstrap, preset="shiny")
+
+        defaults: dict[str, YamlScalarType] = {}
+
+        d_bootstrap = BrandBootstrapConfig._brand_defaults_bootstrap(brand)
+        d_shiny = BrandBootstrapConfig._brand_defaults_shiny(brand)
```

Comment:
Both of these methods could be passed just `brand.defaults`. I also feel like in this case, it's a bit cleaner to drop the `@staticmethod`s and just insert the relevant logic here.

Reply:
Thanks, I tried both suggestions. You're right that these methods could take `brand.defaults` and the logic could be inlined. But I don't think it's cleaner when they're separated and I think `brand` is the right entry point. Taking both suggestions makes the `.from_brand()` method a bit messier and the mess makes it harder to track how the defaults are handled.

Reply:
This feels cleaner to me

[theme-brand-diff.patch](https://github.com/user-attachments/files/17523843/theme-brand-diff.patch)